jplag/src/test/java/de/jplag/clustering/ClusteringTest.java

```diff
@@ -0,0 +1,267 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.List;
+import java.util.Scanner;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.commons.math3.linear.Array2DRowRealMatrix;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.mockito.invocation.InvocationOnMock;
+
+import de.jplag.JPlagComparison;
+import de.jplag.Submission;
+import de.jplag.clustering.algorithm.GenericClusteringAlgorithm;
+import de.jplag.clustering.algorithm.SpectralClustering;
+import de.jplag.clustering.preprocessors.CumulativeDistributionFunctionPreprocessor;
+
+public class ClusteringTest {
+
+    @Test
+    public void testClustering() {
+        List<Submission> submissions = IntStream.range(0, 4).mapToObj(x -> mock(Submission.class)).collect(Collectors.toList());
+        List<JPlagComparison> comparisons = new ArrayList<>(6);
+        for (int i = 0; i < submissions.size(); i++) {
+            for (int j = i + 1; j < submissions.size(); j++) {
+                JPlagComparison comparison = mock(JPlagComparison.class);
+                when(comparison.getFirstSubmission()).thenReturn(submissions.get(i));
+                when(comparison.getSecondSubmission()).thenReturn(submissions.get(j));
+                comparisons.add(comparison);
+            }
+        }
+
+        // Mock algorithm that returns everything in a single cluster
+        GenericClusteringAlgorithm algorithm = mock(GenericClusteringAlgorithm.class);
+        when(algorithm.cluster(any(RealMatrix.class))).then((InvocationOnMock invocation) -> {
+            RealMatrix arg = invocation.getArgument(0);
+            Collection<Collection<Integer>> result = List.of(IntStream.range(0, arg.getRowDimension()).boxed().collect(Collectors.toList()));
+            return result;
+        });
+
+        ClusteringAdapter clustering = new ClusteringAdapter(comparisons, x -> 0.f);
+        ClusteringResult<Submission> clusteringResult = clustering.doClustering(algorithm);
+
+        Collection<Collection<Submission>> expectedResult = List.of(submissions);
+
+        assertEquals(expectedResult, clusteringResult.getClusters().stream().map(Cluster::getMembers).collect(Collectors.toList()));
+    }
+
+    private String str(float f) {
+        return String.format("%.4f", f);
+    }
+
+    private float harmonicMean(float a, float b) {
+        return 2 * a * b / (a + b);
+    }
+
+    private URL loadFromClasspath(String file) throws FileNotFoundException {
+        URL url = getClass().getClassLoader().getResource(file);
+        if (url == null) {
+            throw new FileNotFoundException(file + " not found. Was the PseudonymizedReports submodule initialized?");
+        }
+        return url;
+    }
+
+    @Test
+    @Ignore
+    public void aClusteringOld() throws FileNotFoundException, URISyntaxException {
```

Comment:
I am a bit confused by these ignored tests. Obviously, those were designed for the pseudonymized reports. Leaving them in for us to use is fine, but they should be well documented. Also, they should be separate from normal unit tests like the non-descriptive `testClustering` test a few lines above. Please refer to the [VolumeTest](https://github.com/jplag/JPlag/blob/aab5927fa8e44c1eee2bb575790096c64316493c/jplag/src/test/java/de/jplag/VolumeTest.java) class to see what I mean by documented and separate.

Moreover, I fail to see any assertions. A test case without assertions is not a test case. Also, without automated checking, this test case has little to no value for future maintenance. I would be important that at least some result checking is done. Else, we can remove these tests.

Reply:
You might not want to use Ignore. 

Simply add an assume statement (JUnit Assume) that checks whether the neccessary files are present.